clients/drcachesim/tests/snoop_filter_test.cpp

```diff
@@ -0,0 +1,205 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+// Unit tests for snoop filter.
+#include <iostream>
+#undef NDEBUG
+#include <assert.h>
+#include "snoop_filter_test.h"
+#include "simulator/cache_lru.h"
+#include "snoop_filter.h"
+#include "cache.h"
+#include <algorithm>
+#include <memory>
+
+enum {
+    CPU_0,
+    CPU_1,
+    CPU_2,
+    CPU_3,
+};
+
+static const addr_t ADDR_A = 128;
+static const int SNOOPED_FILTER_LINE_SIZE = 32;
+
+class snoop_filter_test_t {
+
+public:
+    snoop_filter_test_t(const int num_cores)
+        : num_cores_(num_cores)
+    {
+    }
+    ~snoop_filter_test_t()
+    {
+        if (snooped_caches_ != NULL) {
+            delete[] snooped_caches_;
+        }
+        if (snoop_filter_ != NULL) {
+            delete snoop_filter_;
+        }
+    }
+
+    void
+    initialize_caches_and_snoop_filter()
+    {
+        cache_t *llc = new cache_lru_t;
+        if (!llc->init(/*associatibity=*/8, /*line_size=*/64,
+                       /*total_size=*/1024, nullptr,
+                       new cache_stats_t((int)64, "", true))) {
+            std::string error_string =
+                "Usage error: failed to initialize LL cache.  Ensure sizes and "
+                "associativity are powers of 2, that the total size is a multiple "
+                "of the line size, and that any miss file path is writable.";
+            std::cerr << error_string << "\n";
+            return;
+        }
+
+        unsigned int total_snooped_caches = 1 * num_cores_;
+        snooped_caches_ = new cache_t *[total_snooped_caches];
+        for (int i = 0; i < num_cores_; i++) {
+            snooped_caches_[i] = new cache_lru_t;
+            snooped_caches_[i]->init(
+                /*associativity=*/4, SNOOPED_FILTER_LINE_SIZE,
+                /*total_size=*/256, llc,
+                new cache_stats_t((int)SNOOPED_FILTER_LINE_SIZE, "", true, true),
+                /*prefetcher=*/nullptr,
+                /*inclusive=*/true, /*coherence_cache=*/true, i, snoop_filter_);
+        }
+
+        if (!snoop_filter_->init(snooped_caches_, num_cores_)) {
+            std::cerr << "Usage error: failed to initialize snoop filter.\n";
+            return;
+        }
+    }
```

Comment:
Looks like `llc` and its stats class are *always* leaked.  (The parent pointer stored in each snooped_caches_ entry is not freed.)